src/Google.Events.Protobuf.Tests/ProtobufJsonCloudEventFormatterTest.cs

```diff
@@ -0,0 +1,240 @@
+
+            Assert.Equal(2, actualEvents.Count);
+            AssertCloudEventsEqual(expectedEvent1, actualEvents[0]);
+            AssertCloudEventsEqual(expectedEvent2, actualEvents[1]);
+        }
+
+        [Fact]
+        public void EncodeStructured()
+        {
+            var cloudEvent = CreateSampleEvent();
+            var expectedBytes = TestResourceHelper.LoadBytes("structured-mode-body.json");
+
+            var converter = new ProtobufJsonCloudEventFormatter<StorageObjectData>();
+            var actualBytes = converter.EncodeStructuredModeMessage(cloudEvent, out var contentType);
+            Assert.Equal("application/cloudevents+json", contentType.MediaType);
+
+            AssertJsonBytesEqual(expectedBytes, actualBytes);
+        }
+
+        [Fact]
+        public void EncodeStructured_NoData()
+        {
+            var cloudEvent = CreateSampleEvent();
+            cloudEvent.Data = null;
+
+            var converter = new ProtobufJsonCloudEventFormatter<StorageObjectData>();
+            var bytes = converter.EncodeStructuredModeMessage(cloudEvent, out var contentType);
+            Assert.Equal("application/cloudevents+json", contentType.MediaType);
+
+            var obj = JObject.Parse(Encoding.UTF8.GetString(bytes));
+            Assert.False(obj.ContainsKey("data"));
+        }
+
+        [Fact]
+        public void EncodeBinary()
+        {
+            var cloudEvent = CreateSampleEvent();
+            var expectedBytes = TestResourceHelper.LoadBytes("binary-mode-body.json");
+
+            var converter = new ProtobufJsonCloudEventFormatter<StorageObjectData>();
+            var actualBytes = converter.EncodeBinaryModeEventData(cloudEvent);
+
+            AssertJsonBytesEqual(expectedBytes, actualBytes);
+        }
+
+        [Fact]
+        public void EncodeBinary_NoData()
+        {
+            var cloudEvent = CreateSampleEvent();
+            cloudEvent.Data = null;
+
+            var converter = new ProtobufJsonCloudEventFormatter<StorageObjectData>();
+            Assert.Empty(converter.EncodeBinaryModeEventData(cloudEvent));
+        }
+
+        [Fact]
+        public void EncodeBatch()
+        {
+            var batch = new[] { CreateSampleEvent(), CreateSampleEvent2() };
+            var expectedBytes = TestResourceHelper.LoadBytes("batch-body.json");
+
+            var converter = new ProtobufJsonCloudEventFormatter<StorageObjectData>();
+            var actualBytes = converter.EncodeBatchModeMessage(batch, out var contentType);
+            Assert.Equal("application/cloudevents-batch+json", contentType.MediaType);
+
+            AssertJsonBytesEqual(expectedBytes, actualBytes);
+        }
+
+        private static void AssertJsonBytesEqual(byte[] expected, byte[] actual)
+        {
+            JToken expectedObject = Parse(expected);
+            JToken actualObject = Parse(actual);
+            Assert.Equal(expectedObject, actualObject);
+
+            JToken Parse(byte[] bytes)
```

Comment:
Is this why you took the Newtonsoft dependency right? There's no easy way to do this with System.Json? Just curious mostly.

Reply:
Json.NET is definitely easier to use for this sort of thing, yes. It's just a bigger API with more convenience methods etc.